### ISSUE TITLE
NEW: Built-in Silencing 

### ISSUE THREAD

**Issue by calebhailey** [sensu/sensu#1328]
First class support for silencing is coming to Sensu Core! Silencing 2.0 will introduce a new, dedicated `/silence` API endpoint and specification. By building Sensu's new built-in silencing on Sensu subscriptions, Sensu will provide support for silencing events for an entire infrastructure, all results of specific checks, and/or a specific client (see: #1327). 

## `/silence (POST)` API

To update the silenced registry, post JSON data to the new `/silence` API.

For example, to silence events for clients with the subscription "load-balancer":

~~~ json 
{
  "subscription": "load-balancer",
  "expires": 1468881576,
  "reason": "scheduled maintenance",
  "creator": "portertech"
}
~~~

For example, to silence events for the check "haproxy-backends` when the client has the subscription "load-balancer":

~~~ json 
{
  "subscription": "load-balancer",
  "check": "haproxy-backends",
  "expires": 1468881576,
  "reason": "scheduled maintenance",
  "creator": "portertech"
}
~~~

For example, to silence events for the client "i-424242", leveraging its automatic per-client subscription https://github.com/sensu/sensu/issues/1327:

~~~ json 
{
  "subscription": "client:i-424242",
  "expires": 1468881576,
  "reason": "troubleshooting issue",
  "creator": "portertech"
}
~~~

For example, to silence events for the check "check-cron":

~~~ json 
{
  "check": "check-cron",
  "expires": 1468881576,
  "reason": "troubleshooting issue",
  "creator": "portertech"
}
~~~

## `/silenced (GET)` API 

Query the silenced registry: 

~~~ json
[
  {
    "subscription": "load-balancer",
    "check": "haproxy-backends",
    "expires": 1468881576,
    "reason": "scheduled maintenance",
    "creator": "portertech"
  },
  {
    "subscription": "client:i-424242",
    "expires": 1468881576,
    "reason": "troubleshooting issue",
    "creator": "portertech"
  },
  {
    "check": "check-cron",
    "expires": 1468881576,
    "reason": "troubleshooting issue",
    "creator": "portertech"
  }
]
~~~

## `/silenced/subscriptions/:subscription (GET)` API

Fetch a specific entry from the silenced registry, per subscription (e.g. `load-balancer`): 

~~~ shell
$ curl -s http://localhost:5678/silenced/subscriptions/load-balancer
[
  {
    "subscription": "load-balancer",
    "check": "haproxy-backends",
    "expires": 1468881576,
    "reason": "scheduled maintenance",
    "creator": "portertech"
  }
]
~~~

## `/silenced/checks/:check (GET)` API

Fetch a specific entry from the silenced registry, per check (e.g. `haproxy-backends`): 

~~~ shell
$ curl -s http://localhost:5678/silenced/checks/haproxy-backends
[
  {
    "subscription": "load-balancer",
    "check": "haproxy-backends",
    "expires": 1468881576,
    "reason": "scheduled maintenance",
    "creator": "portertech"
  }
]
~~~

EDIT: to clarify, this behavior is going to being implemented in Sensu Core proper, which will supercede the legacy support for "silence stashes" in the sensu-plugin gem (though, support for silence stashes won't be immediately deprecated, to allow time for the community & plugin ecosystem to migrate to the new solution). Furthermore, a new handler definition attribute will be implemented to allow certain handlers to be executed even if the event has been silenced (i.e. overriding the silence).

**Comment by quodlibetor** [comment]
Is this going to operate the same way as the current silence stash API -- where handler scripts are responsible for checking for themselves if they are silenced -- or is Sensu going to simply not run handlers for silenced subscriptions?

**Comment by calebhailey** [comment]
@quodlibetor this is going to be implemented in Sensu Core (in the server), in which case handlers won't even be run for silenced events. We will also be providing a handler definition attribute to override this behavior (for handlers that you want to run regardless of silencing), which I should have mentioned above. I'll edit my description to explain this.

**Comment by quodlibetor** [comment]
That's fantastic news! It'll make it much easier to write handlers in non-ruby without needing to port sensu-plugin. People who've written custom handlers will probably need to be especially careful on this update, though.

**Comment by cwjohnston** [comment]
@portertech @calebhailey any thoughts about acknowledgement behavior as described in sensu-plugins/sensu-plugin#135? Seems like it could become in-scope for the silencing implementation described here.

**Comment by fessyfoo** [comment]
I think this is missing a way to silence by client + check name ? 

sensu-handler in https://github.com/sensu-plugins/sensu-plugin currently checks the stash api for three types of silenced:

1. by client `/silence/:client`
2. by client and checkname `/slience/:client/:check`
3. by check name `/silence/all/:check`

this proposal introduces a 4th

4. by subscription

and then #1327 maps by client into by subscription via automatic per client subscriptions.  handling the 'by client'  type silence.   

does that also include mapping per check name  and per client + check name to their own subscriptions as well?   otherwise 'by client + check' and 'by check name' appear left out?

**Comment by fessyfoo** [comment]
I propose that the implementation also support "acknowledgements"  described here https://github.com/sensu-plugins/sensu-plugin/pull/135  

this would require not filtering resolution events when a silence is of type acknowledgment as well as deleting the silence when the resolution event happens.

**Comment by portertech** [comment]
@cwjohnston for the logic described in https://github.com/sensu-plugins/sensu-plugin/pull/135, I believe we can add an optional attribute to this proposed silencing functionality instead of using a "type", for example `"expire_on_resolve": true`.

**Comment by portertech** [comment]
@fessyfoo with the introduction of automatic per-client unique subscriptions, https://github.com/sensu/sensu/issues/1327, there's no need for `/silence/:client`, as it is the same as `/silence/:subscription` where the subscription is the client name. The examples in the proposed spec above missed the very important silence attribute `check`, e.g. `"check": "check-http"`.

**Comment by portertech** [comment]
@RyanJarv perhaps we could add a silence attribute like `"severities": ["warning"]` to accomplished the desired effect explained in #940. I need to first experiment and test the impact it could have on performance etc.

**Comment by portertech** [comment]
@calebhailey I am working to implement the spec described in this issue, however, it may grow to include attribute/features like `expire_on_resolve` and `severities`, either for 0.26 or a future release :+1:

**Comment by portertech** [comment]
I am altering the proposal description to more accurately represent the API.

**Comment by fessyfoo** [comment]
got it,  `/silence/:subscription` where `:subscription` is the [automatic client subscription](https://github.com/sensu/sensu/issues/1327), and  the silence has attribute `check: :check_name` indeed cover the `client + check` type silence.     thanks for the clarification.  (and proposal update. ) 

what about the 'by check'  silence,   sensu-plugin's `/silence/all/:check`,  admittedly I've only used that once.  but  i did find someone else who needed to do that: https://github.com/sensu/uchiwa/issues/458#issuecomment-228486642

**Comment by portertech** [comment]
@fessyfoo I have updated the proposal description to include the silence `"check"` post body attribute.

**Comment by fessyfoo** [comment]
perhaps this for silence by 'check name' ? 

to silence all checks of a particular name across all subscriptions: 

```
{
  "subscription": "all",
  "check": "haproxy-backends",
  "expires": 1468881576,
  "reason": "check is broken every where we rolled it out.",
  "creator": "portertech"
}

```

**Comment by portertech** [comment]
@fessyfoo I usually have an "all" or "production" subscription for clients, so there's really no need for the special exception.

**Comment by portertech** [comment]
I would prefer to not include a "subscription" in the post body, but then the proposed GET API breaks.

**Comment by portertech** [comment]
Updating the proposed solution again :+1:

**Comment by portertech** [comment]
@fessyfoo I have updated the proposal, WDYT? Supports silencing a whole subscription, a subscription + check, and just a check.

**Comment by fessyfoo** [comment]
> @fessyfoo I have updated the proposal, WDYT? Supports silencing a whole subscription, a subscription + check, and just a check.

looks great.  feels fully backwards compatible with what's in sensu-plugin.

**Comment by fessyfoo** [comment]
> @fessyfoo I usually have an "all" or "production" subscription for clients, so there's really no need for the special exception.

On thinking it over though.  I feel that having an "all" subscription is a pretty acceptable pattern for the silence by check use case.    and using that keeps the api a lot simpler.      maybe an automatic "all"  subscription just like the automatic client subscriptions is a better approach?

**Comment by portertech** [comment]
@fessyfoo supporting `/silenced/checks/:check` does however allow users to see all (not just "all") subscriptions that have a check silenced.

**Comment by fessyfoo** [comment]
would `GET /silenced/checks/:check`  return multiple results  if the same check name is silenced under multiple subscriptions?   or some kind of merged result?

**Comment by portertech** [comment]
@fessyfoo it would return an array of silenced registry entries for that check.

**Comment by fessyfoo** [comment]
@portertech  we're trying to plan which way we should go regarding acknowledgments `expire_on_resolve`  any clues on timing for the release of built in silencing and subsequent support of `expire_on_resolve` ?   our other alternative is to run with a fork of sensu-plugin that [includes support for expire_on_resolve](https://github.com/sensu-plugins/sensu-plugin/pull/135).  since that's not the long term direction for silencing we'd rather not go down that path.

**Comment by portertech** [comment]
I opened a WIP pull-request for an implementation of this feature https://github.com/sensu/sensu/pull/1400

**Comment by fessyfoo** [comment]
consider https://github.com/sensu/uchiwa/issues/523 is essentially requesting a "silence all"  feature.

**Comment by portertech** [comment]
@fessyfoo please see @cwjohnston's comment https://github.com/sensu/uchiwa/issues/523#issuecomment-237969377

**Comment by portertech** [comment]
Closing this, as https://github.com/sensu/sensu/pull/1400 has been merged into the 0.26 release :+1: 

The implementation also included `expire_on_resolve` to satisfy the desire for "acknowledgements" :-)

@fessyfoo ^^
